edb/pgsql/resolver/resolver.py

```diff
@@ -0,0 +1,645 @@
+from edb.pgsql import common as pgcommon
+from edb.pgsql.compiler import astutils as pgastutils
+
+from edb.schema import objtypes as s_objtypes
+from edb.schema import links as s_links
+
+from . import dispatch
+from . import context
+
+Context = context.ResolverContextLevel
+
+
+@dispatch._resolve.register
+def resolve_SelectStmt(
+    stmt: pgast.SelectStmt, *, ctx: Context
+) -> pgast.SelectStmt:
+
+    # UNION
+    if stmt.larg:
+        with ctx.isolated() as subctx:
+            stmt.larg = dispatch.resolve(stmt.larg, ctx=subctx)
+    if stmt.rarg:
+        with ctx.isolated() as subctx:
+            stmt.rarg = dispatch.resolve(stmt.rarg, ctx=subctx)
+    if stmt.larg or stmt.rarg:
+        return stmt
+
+    # CTEs
+    ctes = None
+    if stmt.ctes:
+        ctes = [dispatch.resolve(cte, ctx=ctx) for cte in stmt.ctes]
+
+    # FROM
+    from_clause: List[pgast.BaseRangeVar] = []
+    for clause in stmt.from_clause:
+
+        with ctx.empty() as subctx:
+            from_clause.append(dispatch.resolve(clause, ctx=subctx))
+
+            # pull result relation of inner scope into own scope
+            if subctx.scope.rel.columns:
+                ctx.scope.tables.append(subctx.scope.rel)
+            if subctx.scope.join_relations:
+                ctx.scope.tables.extend(subctx.scope.join_relations)
+
+    where = dispatch.resolve_opt(stmt.where_clause, ctx=ctx)
+    target_list = [
+        c for t in stmt.target_list for c in resolve_ResTarget(t, ctx=ctx)
+    ]
+
+    distinct_clause = None
+    if stmt.distinct_clause:
+        distinct_clause = [
+            (c if isinstance(c, pgast.Star) else dispatch.resolve(c, ctx=ctx))
+            for c in stmt.distinct_clause
+        ]
+
+    sort_clause = dispatch.resolve_opt_list(stmt.sort_clause, ctx=ctx)
+    limit_offset = dispatch.resolve_opt(stmt.limit_offset, ctx=ctx)
+    limit_count = dispatch.resolve_opt(stmt.limit_count, ctx=ctx)
+
+    res = pgast.SelectStmt(
+        distinct_clause=distinct_clause,
+        from_clause=from_clause,
+        target_list=target_list,
+        where_clause=where,
+        sort_clause=sort_clause,
+        limit_offset=limit_offset,
+        limit_count=limit_count,
+        ctes=ctes,
+    )
+    return res
+
+
+@dispatch._resolve.register
+def resolve_DMLQuery(cte: pgast.DMLQuery, *, ctx: Context) -> pgast.DMLQuery:
+    raise errors.UnsupportedFeatureError(
+        'DML queries (INSERT/UPDATE/DELETE) are not supported'
+    )
+
+
+@dispatch._resolve.register
+def resolve_CommonTableExpr(
+    cte: pgast.CommonTableExpr, *, ctx: Context
+) -> pgast.CommonTableExpr:
+    with ctx.isolated() as subctx:
+
+        query = dispatch.resolve(cte.query, ctx=subctx)
+
+        result = context.CTE()
+        result.name = cte.name
+        result.columns = []
+
+        cols_query = subctx.scope.rel.columns
+        cols_names = cte.aliascolnames
+
+        if cols_names and len(cols_query) != len(cols_names):
+            raise errors.QueryError(
```

Comment:
I've included the context, but it doesn't work that well, because it reports location within the SQL query, not the whole EdgeQL query:

```
error: QueryError: unknown object `MessaGe`
  ┌─ query:1:15
  │
1 │ describe alter 'SELECT * FROM "messaGe"';
  │               ^ error
```

But this will work ok, when SQL will be parsed directly.